### PR TITLE
Update std.traits to reflect renaming of TypeTuple to AliasSeq

### PR DESCRIPTION
**Pull request by wilzbach**
Revived #3756 as is (just rebased to latest Master).

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -248,44 +248,33 @@ private
...
-/// ditto.
 template ConstOf(T)       { alias ConstOf       =        const(T) ; }
-/// ditto.
 template SharedOf(T)      { alias SharedOf      =       shared(T) ; }
-/// ditto.
 template SharedInoutOf(T) { alias SharedInoutOf = shared(inout(T)); }
-/// ditto.
 template SharedConstOf(T) { alias SharedConstOf = shared(const(T)); }
-/// ditto.
 template ImmutableOf(T)   { alias ImmutableOf   =    immutable(T) ; }
```
Comment (wilzbach) [inline std/traits.d]: does anyone know what @jamadagni was trying to do here?
AFAICT this change undocuments all these templates, doesn't it?

**Comment by wilzbach** [comment]
So I took #3756 as is (well apart from the necessary changes during the rebase) and started to amend my changes. The two important ones:
- fixed compilation (thanks @tsbockman)
- added deprecation messages and a specific date to the `deprecated alias`

**Comment by andralex** [comment]
dat copyright year update

**Comment by andralex** [comment]
What's Jenkins unhappy about?

**Comment by wilzbach** [comment]
> dat copyright year update

There are (a lot) more changes - you need to click on "Load diff" ;-)

> What's Jenkins unhappy about?

Martin just started the project tester. However many of the projects are still broken with the current DMD / Phobos. So please ignore it for now.

File: std/traits.d [hunk 2]
```diff
@@ -64,58 +64,58 @@
...
+ *           $(LREF isAggregateType)
+ *           $(LREF isSomeString)
+ *           $(LREF isNarrowString)
+ *           $(LREF isAutodecodableString)
+ *           $(LREF isPointer)
  * ))
  * $(TR $(TD Type behaviours) $(TD
  *           $(LREF isAbstractClass)
  *           $(LREF isAbstractFunction)
+ *           $(LREF isInstanceOf)
```
Comment (andralex) [inline std/traits.d]: why did you move this? I think it was ordered alphabetically

File: std/traits.d [hunk 3]
```diff
@@ -64,58 +64,58 @@
...
  *           $(LREF isCallable)
  *           $(LREF isDelegate)
- *           $(LREF isExpressions)
  *           $(LREF isFinalClass)
  *           $(LREF isFinalFunction)
  *           $(LREF isFunctionPointer)
- *           $(LREF isInstanceOf)
+ *           $(LREF isSomeFunction)
```
Comment (andralex) [inline std/traits.d]: keep ordered

File: std/traits.d [hunk 4]
```diff
@@ -124,8 +124,8 @@
  *           $(LREF mostNegative)
  *           $(LREF OriginalType)
  *           $(LREF PointerTarget)
- *           $(LREF Signed)
  *           $(LREF Unqual)
+ *           $(LREF Signed)
```
Comment (andralex) [inline std/traits.d]: this wasn't ordered but probably should

File: std/traits.d [hunk 5]
```diff
@@ -146,7 +146,7 @@
  * Copyright: Copyright Digital Mars 2005 - 2009.
  * License:   $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
```
Comment (andralex) [inline std/traits.d]: yum

File: std/traits.d [hunk 6]
```diff
@@ -146,7 +146,7 @@
...
  * Authors:   $(HTTP digitalmars.com, Walter Bright),
- *            Tomasz Stachowiak ($(D isExpressions)),
+ *            Tomasz Stachowiak ($(D isExpressionSeq)),
```
Comment (andralex) [inline std/traits.d]: Use backticks since you touched it?

File: std/traits.d (hunk 1 above)
Comment (andralex) [inline std/traits.d]: I'd say undo this change

File: std/traits.d [hunk 7]
```diff
@@ -924,65 +910,67 @@ template ReturnType(func...)
...
 
+
 /***
-Get, as a tuple, the types of the parameters to a function, a pointer
-to function, a delegate, a struct with an $(D opCall), a pointer to a
-struct with an $(D opCall), or a class with an $(D opCall).
+Get an `AliasSeq` of the types of the parameters to a callable.
 */
-template Parameters(func...)
+template ParameterTypes(func...)
```
Comment (andralex) [inline std/traits.d]: Is this a breaking change?

File: std/traits.d [hunk 8]
```diff
@@ -1039,51 +1027,48 @@ template ParameterStorageClassTuple(func...)
...
+    static assert(PSC_func.length == 3); // three parameters
+    static assert(PSC_func[0] == PSC.ref_);
+    static assert(PSC_func[1] == PSC.out_);
+    static assert(PSC_func[2] == PSC.none);
 }
 
 @safe unittest
 {
-    alias STC = ParameterStorageClass;
+    alias PSC = ParameterStorageClass;
```
Comment (andralex) [inline std/traits.d]: this kind of thing seems gratuitous

File: std/traits.d [hunk 9]
```diff
@@ -924,65 +910,67 @@ template ReturnType(func...)
...
         static assert(0, "argument has no parameters");
 }
 
+// @@@DEPRECATED_2017-12@@@
+deprecated("It will be removed from Phobos in December 2017. Use ParameterTypes")
+alias ParameterTypeTuple = ParameterTypes;
+
+// @@@DEPRECATED_2017-12@@@
+deprecated("It will be removed from Phobos in December 2017. Use ParameterTypes")
+alias Parameters = ParameterTypes;
```
Comment (andralex) [inline std/traits.d]: this seems gratuitous

File: std/traits.d [hunk 10]
```diff
@@ -924,65 +910,67 @@ template ReturnType(func...)
...
     static if (is(FunctionTypeOf!func P == function))
-        alias Parameters = P;
+        alias ParameterTypes = P;
     else
         static assert(0, "argument has no parameters");
 }
 
+// @@@DEPRECATED_2017-12@@@
+deprecated("It will be removed from Phobos in December 2017. Use ParameterTypes")
+alias ParameterTypeTuple = ParameterTypes;
```
Comment (andralex) [inline std/traits.d]: this seems something we can live with

**Comment by wilzbach** [comment]
> This seems to be a pass that someone made through a personal project after finding a nicer naming convention. I think it breaks too many things for too little benefit.

> I think some of the things can be cherry picked, e.g. name of nodes in the grammar. Looks like a long task, thanks for taking it.

Okay started with the least controversial bit: triggering the deprecation warnings for `std.typetuple` -> #4968

**Comment by andralex** [comment]
@MartinNowak - I think you should look at this.

**Comment by wilzbach** [comment]
Closing this for the same reason as #4968 (not much time, no progress / consensus in sight).
However, as mentioned on the other PR I am still interested in seeing this happening, so @ D enthusiasts from the future please don't take this as a sign, that we will never clean up the mess `std.typetuple`, just don't give up trying ;-)
